File: go-version/leetcode/LC2275/largest_combination.go
package lc2275

import "math/bits"

func largestCombination(candidates []int) int {
	//取出数组中的最大值
	mx := candidates[0]
	for i := range candidates {
		mx = max(mx, candidates[i])
	}
	//最大值二进制有几位
	maxLen := bits.Len(uint(mx))

	//枚举二进制位，计算该位非零的条件下，最多可以取出多少个数
	//求出满足条件的最大值
	ans := 0
	for i := 0; i < maxLen; i++ {
		curCnt := 0
		for _, num := range candidates {
			curCnt += num >> i & 1
		}
		ans = max(ans, curCnt)
	}
	return ans
}

File: python-version/LC2275/largest_combination.py
import unittest

class Solution:
    def largestCombination(self, candidates: list[int]) -> int:
        # 求出数组的最大值
        mx = max(candidates)
        # 计算最大值的二进制位数
        maxLen = mx.bit_length()

        # 枚举二进制位，计算满足条件的数字个数
        ans = 0
        for i in range(maxLen):
            curCnt = 0
            for num in candidates:
                curCnt += num >> i & 1
            ans = max(ans, curCnt)
        return ans
    
class SolutionTest(unittest.TestCase):
    def TestLargestCombination(self):
        solution = Solution()
        self.assertEqual(4, solution.largestCombination([16,17,71,62,12,24,14]))
        self.assertEqual(2, solution.largestCombination([8, 8]))


if __name__ == '__main__':
    unittest.main()